watchers/configmap_watcher.go

```diff
@@ -68,21 +67,21 @@ func checkConfigMapChanges(eventChannel <-chan watch.Event, dataField string, da
 	}
 }
 
-func createWatcher(clientset kubernetes.Interface, namespace string, configMapName string, watcher *watch.Interface) error {
-	err := retry.OnError(backoff, retriable, func() error {
-		var innerErr error
-		*watcher, innerErr = clientset.CoreV1().ConfigMaps(namespace).Watch(context.TODO(),
-			metav1.SingleObject(metav1.ObjectMeta{Name: configMapName, Namespace: namespace}))
-		if innerErr != nil {
-			logger.Info("cannot create watcher", "namespace", namespace, "configMap name", configMapName, "error", innerErr)
-		}
-		return innerErr
-	})
-	return err
+// watcher implements config.Watch interface which will create a ConfigMapWatcher.
+type watcher struct {
+	clientset     kubernetes.Interface
+	namespace     string
+	configMapName string
 }
 
-func retriable(err error) bool {
-	retry := err != nil
-	logger.V(1).Info("cannot create watcher", "retriable", retry, "error", err)
-	return retry
+func (w watcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	watcher, err := w.clientset.CoreV1().ConfigMaps(w.namespace).Watch(context.Background(),
+		metav1.ListOptions{
+			ResourceVersion:     options.ResourceVersion,
+			AllowWatchBookmarks: options.AllowWatchBookmarks,
```

Comment:
I don't know how exactly bookmarks work (I red [this](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks) ).
Last sentence says: 
> but you shouldn't assume bookmarks are returned at any specific interval, nor can clients assume that the API server will send any BOOKMARK event even when requested

Should we take care of this? (Asking to better understand the code)

Reply:
a little bit of context here:
the whole process changed with the RetryWatcher. Now the RetryWatcher wants a structure with some sort of creator watcher method which is called by the RetryWatcher when it starts. If you look at the code of the RetryWatcher (https://github.com/kubernetes/client-go/blob/28ccde769fc5519dd84e5512ebf303ac86ef9d7c/tools/watch/retrywatcher.go#L100) is like:
```go
func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
	watcher, err := rw.watcherClient.Watch(metav1.ListOptions{
		ResourceVersion:     rw.lastResourceVersion,
		AllowWatchBookmarks: true,
	})
```
So, what I've done is take these arguments and inject them along with my configMap name into the `ListOptions` object passed to `configMapWatcher`. In this context, the `AllowWatchBookmarks` is always true.

Reply:
Thanks!